### PR TITLE
Allow more fine grained control of output format

### PR DESCRIPTION
**Pull request by wlandry**
This adds the options

- --null
- --[no--]with-filename
- --[no--]line-number

to allow more fine grained control of the output.  It also regularizes the tty behavior by making it set these options, rather than checking for a tty and then overriding behavior at the last moment.

There is also a bit of cleanup of unused enums and a fix to includes.

Default output should remain the same.

Please let me know if this is the kind of thing you want, or if you rather go in a different direction.

### CODE REVIEW

**Comment by gvansickle** [comment]
Hi @wlandry , thanks for the PR!  Taking a look at it now.

**Review by gvansickle — CHANGES_REQUESTED** [review]
Yeah, I definitely like the direction of these changes @wlandry .  See the couple minor review comments.

File: src/ArgParse.cpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright 2015-2022 Gary R. Van Sickle (grvs@users.sourceforge.net).
+ * Copyright (C) 2022 Free Software Foundation, Inc.
```
[the remark this answers is not in the document]
Comment (wlandry) [reply src/ArgParse.cpp:3]: This is intentional.  I lifted the help text directly from GNU grep.

File: src/ArgParse.cpp [hunk 2]
```diff
@@ -524,11 +523,22 @@ static const std::array f_raw_options = std::to_array<PreDescriptor>({
 		{ OPT_WORDREGEX, 0, "w", "word-regexp", Arg::None, "PATTERN must match a complete word."},
 		{ OPT_LITERAL, 0, "Q", "literal", Arg::None, "Treat all characters in PATTERN as literal."},
 	{ "Search Output:" },
+		{ OPT_PREFIX_FILE, ENABLE, "H", "with-filename", Arg::None, "Prefix the result with the file name."},
+		{ OPT_PREFIX_FILE, DISABLE, "h", "no-with-filename", Arg::None, "Prefix the file name on a separate line before matches (default)."},
+		{ OPT_LINE_NUMBER, ENABLE, "n", "line-number", Arg::None, "Print the line number of each match (default)."},
```
[the remark this answers is not in the document]
Comment (wlandry) [reply src/ArgParse.cpp:528]: I removed the -n short option from --no-recurse.

File: src/ArgParse.h [hunk 3]
```diff
@@ -123,6 +123,11 @@ class ArgParse
 	bool m_pattern_is_literal { false };
 
 	/// true if we should print the column of the first match after the line number.
+	bool m_line_number { true };
+
+	/// true if we should print the column of the first match
+	/// after the line number.  Setting this to true implies will
+	/// also set m_line to true.
```
[the remark this answers is not in the document]
Comment (wlandry) [reply src/ArgParse.h]: Fixed

File: src/MatchList.cpp [hunk 4]
```diff
@@ -74,86 +74,57 @@ void MatchList::Print(std::ostream &sstrm, OutputContext &output_context) const
 
 	std::string composition_buffer;
 	composition_buffer.reserve(256);
-
-	// The only real difference between TTY vs. non-TTY printing here is that for TTY we print:
```
[the remark this answers is not in the document]
Comment (wlandry) [reply src/MatchList.cpp:78]: I put the comment in ArgParse.cpp where I check for tty.

File: src/OutputTask.cpp [hunk 5]
```diff
@@ -31,28 +31,14 @@
...
+                       bool flag_column, bool flag_nullsep, sync_queue<MatchList> &input_queue)
+  : m_input_queue(input_queue), m_enable_color(flag_color),
+    m_prefix_file(flag_prefix_file), m_print_line_number(flag_line_number), 
+    m_print_column(flag_column), m_nullsep(flag_nullsep)
 {
-	// Determine if the output is going to a terminal.  If so we'll use color by default, group the matches under
-	// the filename, etc.
-	m_output_is_tty = isatty(fileno(stdout));
-
-	// Determine whether to enable color or not.
```
[the remark this answers is not in the document]
Comment (wlandry) [reply src/OutputTask.cpp:41]: I incorporated this comment in ArgParse.cpp where I check for tty.

**Comment by wlandry** [comment]
I rebased and push the fixed version.

File: src/ArgParse.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (gvansickle) [reply src/ArgParse.cpp:3]: Ok, looks like you only took the "man grep" text regarding --null from grep, I don't see anything else added from grep.  I hate to say it, but I think this should be changed in one, maybe two ways:
1. Remove the copyright line.  Per GNU/FSF themselves, it's not needed even if you're doing a direct copy/paste, see https://www.gnu.org/licenses/gpl-faq.en.html#AllCompatibility .  We're GPL3 and grep is GPL3+.
2. Shorten/reword the --null help text so it's not a copy/paste.

I'm OK with just # 1, and would merge it, everything else looks good.

**Review by gvansickle — CHANGES_REQUESTED** [review]
Looks great @wlandry .  See my one remaining change request, when that's resolved I'll merge your PR.  Thanks!

**Comment by wlandry** [comment]
I did option 2 and shortened the help text.

**Comment by gvansickle** [comment]
PR merged, thanks @wlandry !

**Pull request merged**